### DISCUSSION TITLE
Can't import a server action from a "use client"; marked file in version 14

### DISCUSSION THREAD

**Discussion by pedro757** [vercel/next.js#57519] [App Router]
It works in version 13.5.6 but 14.0.0 errors

`
Error: You're importing a Client Component ("use client") from another Client Component imported Server Action file ("use server"). This is not allowed due to cyclic module graph between Server and Client.
You can work around it by defining and passing this Server Action from a Server Component into the Client Component via props.
`

Can you improve the error message btw? it's definitely confusing

**Comment by icyJoseph** [comment]
I guess the stability of Server Actions, requires no circular dependencies.

What kind of import chain do you have? I guess you have Client Component module A, that imports from Client Component module B, and B also imports something from A? And at least one of them is using Server Actions? 

Circular dependencies have always been a bit nasty to work with though, you should definitely avoid them.

**Reply by pedro757** [reply]
You're right but IMO this is too complex to keep in mind.

**Reply by icyJoseph** [reply]
This has always been kind of problematic though. 

People typically use linting rules to catch these. https://github.com/import-js/eslint-plugin-import/blob/main/docs/rules/no-cycle.md

**Reply by pedro757** [reply]
Thank you! I will check out that plugin

**Comment by manuganji** [comment]
I also have this issue. It was working in Next@13 but stopped in 14. In my case, I'm passing a validate action to the form and I'm also checking the same validation in a different on submit action. The problem seems to be referring to the validate action both in the form component and in the consumer. There shouldn't be an error because these are 2 separate execution paths and they are not cyclic references.

**Reply by manuganji** [reply]
Also the error is being logged from a completely unrelated component file. It's unhelpful.

**Reply by manuganji** [reply]
Downgraded to 13 for now.
